src/web/demo.rkt

```diff
@@ -336,12 +339,55 @@
                      `(p "You didn't specify a formula (or you specified several). "
                          "Please " (a ([href ,go-back]) "go back") " and try again."))]))
 
+(define (extract-formula-and-seed post-data)
+  ;; Maybe pass in context from the request for the error?
+  (define form-str (hash-ref post-data 'formula (lambda () (raise-herbie-error "Unable to parse formula from ~a\n" post-data))))
+  (define form-seed (hash-ref post-data 'seed (lambda () (raise-herbie-error "Unable to parse seed from ~a\n" post-data))))
+  (list form-str form-seed))
+
+(define (parse-formula-from-string formula-str)
+ (with-handlers ([exn:fail? (lambda (e) 
+  (raise-herbie-error "Unable to parse formula:~a\nCheck for syntax errors and try again.\n" formula-str))])
+  (read-syntax 'web (open-input-string formula-str))))
+
+(define (demo-error go-back e)
+  (response/error
+  "Demo Error"
+  `(div
+    (h1 "Invalid formula")
+    (pre ,(herbie-error->string e))
+    (p
+      "Formula must be a valid program using only the supported functions. "
+      "Please " (a ([href ,go-back]) "go back") " and try again."))))
+
+(define (improve-start-helper req body go-back)
+  (define post-body (request-post-data/raw req))
+  (define post-data (bytes->jsexpr post-body))
+  (match (extract-formula-and-seed post-data)
+    [(list formula-str seed*)
+     ;; TODO Check that seed is a valid seed?
+     (define formula (parse-formula-from-string formula-str))
+     (with-handlers
+         ([exn:fail:user:herbie? (curry demo-error go-back)])
+       (when (eof-object? formula)
+         (raise-herbie-error "no formula specified"))
+       (parse-test formula)
+       (define hash (sha1 (open-input-string (~s 'improve formula (get-seed)))))
+       (body hash formula seed*))]
+    [_
```

Comment:
Is this even possible to hit/test now that parsing valid formulas is handled before it?